pkg/cluster/removeprivatednszone.go

```diff
@@ -27,125 +28,85 @@ func (m *manager) removePrivateDNSZone(ctx context.Context) error {
 
 	if len(zones) == 0 {
 		// fix up any clusters that we already upgraded
-		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			dns, err := m.configcli.ConfigV1().DNSes().Get(ctx, "cluster", metav1.GetOptions{})
-			if err != nil {
-				return err
-			}
-
-			if dns.Spec.PrivateZone == nil ||
-				!strings.HasPrefix(strings.ToLower(dns.Spec.PrivateZone.ID), strings.ToLower(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID)) {
-				return nil
-			}
-
-			dns.Spec.PrivateZone = nil
-
-			_, err = m.configcli.ConfigV1().DNSes().Update(ctx, dns, metav1.UpdateOptions{})
-			return err
-		})
-		if err != nil {
+		if err := utilnet.UpdateDNSs(ctx, m.configcli.ConfigV1().DNSes(), resourceGroupID); err != nil {
 			m.log.Print(err)
 		}
-
 		return nil
 	}
 
-	mcps, err := m.mcocli.MachineconfigurationV1().MachineConfigPools().List(ctx, metav1.ListOptions{})
+	sameNumberOfNodesAndMachines, err := m.sameNumberOfNodesAndMachines(ctx)
 	if err != nil {
 		m.log.Print(err)
 		return nil
 	}
 
-	var machineCount int
-	for _, mcp := range mcps.Items {
-		var found bool
-		for _, source := range mcp.Status.Configuration.Source {
-			if source.Name == "99-"+mcp.Name+"-aro-dns" {
-				found = true
-				break
-			}
-		}
-
-		if !found {
-			m.log.Printf("ARO DNS config not found in MCP %s", mcp.Name)
-			return nil
-		}
-
-		if !ready.MachineConfigPoolIsReady(&mcp) {
-			m.log.Printf("MCP %s not ready", mcp.Name)
-			return nil
-		}
-
-		machineCount += int(mcp.Status.MachineCount)
+	if !sameNumberOfNodesAndMachines {
+		return nil
 	}
 
-	nodes, err := m.kubernetescli.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	clusterVersionIsLessThan4_4, err := version.ClusterVersionIsLessThan4_4(ctx, m.configcli)
 	if err != nil {
 		m.log.Print(err)
 		return nil
 	}
 
-	if len(nodes.Items) != machineCount {
-		m.log.Printf("cluster has %d nodes but %d under MCPs, not removing private DNS zone", len(nodes.Items), machineCount)
+	if clusterVersionIsLessThan4_4 {
+		m.log.Printf("cluster version < 4.4, not removing private DNS zone")
 		return nil
 	}
 
-	cv, err := m.configcli.ConfigV1().ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
-	if err != nil {
-		return err
-	}
-	v, err := version.GetClusterVersion(cv)
-	if err != nil {
+	if err = utilnet.UpdateDNSs(ctx, m.configcli.ConfigV1().DNSes(), resourceGroupID); err != nil {
 		m.log.Print(err)
 		return nil
 	}
 
-	if v.Lt(version.NewVersion(4, 4)) {
-		// 4.3 uses SRV records for etcd
-		m.log.Printf("cluster version < 4.4, not removing private DNS zone")
+	if err := utilnet.RemoveZones(ctx, m.virtualNetworkLinks, m.privateZones, zones, resourceGroup); err != nil {
+		m.log.Print(err)
 		return nil
 	}
+	return nil
+}
```

Comment:
the tests around this function in `pkg/cluster/removeprivatednszone_test.go` had to do its mocking with `mock_privatedns` I'd expect with this type of refactoring the tests to be changed and added to `pkg/util/net.go` and the mocks here to be updated to capture the new use of `utilnet`.